components/script_bindings/str.rs

```diff
@@ -293,6 +299,65 @@ impl DOMString {
     }
 }
 
+/// Because this converts to a DOMString it becomes UTF-8 encoded which is closer to
+/// the spec definition of <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-json-bytes>
+/// but we generally do not operate on anything that is truly a WTF-16 string.
+///
+/// <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-a-json-string>
+pub fn serialize_jsval_to_json_utf8(
+    cx: SafeJSContext,
+    data: HandleValue,
+) -> Result<DOMString, Error> {
+    #[repr(C)]
+    struct ToJSONCallbackData {
+        string: OnceCell<String>,
+    }
+
+    let mut out_str = ToJSONCallbackData {
+        string: OnceCell::new(),
+    };
+
+    #[allow(unsafe_code)]
+    unsafe extern "C" fn write_callback(
+        string: *const u16,
+        len: u32,
+        data: *mut std::ffi::c_void,
+    ) -> bool {
+        let data = data as *mut ToJSONCallbackData;
+        let string_chars = slice::from_raw_parts(string, len as usize);
+        // ToJSON promises it will only ever call the write_callback at most once.
+        let _ = (*data).string.set(String::from_utf16_lossy(string_chars));
```

Comment:
This may be true, but maybe this code could be adapted in case the behavior of `ToJSON` ever changes:
```suggestion
        let _ = (*data).string.get_or_init(Default::default).push_str(String::from_utf16_lossy(string_chars));
```

Reply:
This is a bit awkward, `get_or_init` will not provide a mutable reference, we can `string.get_or_init(||String::from_utf16_lossy(string_chars))` though that doesn't seem to really help in handling the case where `ToJSON` changes behavior to call the `write_callback` mutiple times. (In both cases we'd be ignoring the data passed the second time)

Reply:
Oh, I had forgotten about this particularly annoying bit of the `OnceLock` API. Sorry about that! I think you could switch from `OnceLock` to `Option` here and simply do `(*data).string.get_or_insert_with(Default::default).push_str(String::from_utf16_lossy(string_chars));`